File: sdk/java/src/main/java/io/ancora/handles/RuntimeHandle.java
package io.ancora.handles;

import io.ancora.AncorException;
import io.ancora.ffi.AncorErrorCode;
import io.ancora.ffi.AncoraNative;

import java.lang.foreign.*;

/**
 * Manages the lifetime of an opaque AncorRuntime pointer.
 * Calls ancora_free_runtime when closed. AutoCloseable for try-with-resources.
 */
public final class RuntimeHandle implements AutoCloseable {

    private final MemorySegment ptr;
    private volatile boolean closed = false;

    private RuntimeHandle(MemorySegment ptr) {
        this.ptr = ptr;
    }

    /**
     * Allocate a new native runtime via ancora_runtime_new.
     *
     * @throws UnsatisfiedLinkError if the native library was not loaded
     * @throws AncorException if ancora_runtime_new returns a non-OK code
     */
    public static RuntimeHandle create() throws Throwable {
        if (!AncoraNative.AVAILABLE) {
            throw new UnsatisfiedLinkError("ancora_ffi native library not found");
        }
        try (Arena arena = Arena.ofConfined()) {
            MemorySegment outPtr = arena.allocate(ValueLayout.ADDRESS);
            int rc = (int) AncoraNative.RUNTIME_NEW.invokeExact(outPtr);
            if (rc != 0) {
                throw new AncorException(rc, "ancora_runtime_new failed");
            }
            MemorySegment runtimePtr = outPtr.get(ValueLayout.ADDRESS, 0);
            if (runtimePtr.equals(MemorySegment.NULL)) {
                throw new AncorException(AncorErrorCode.INTERNAL, "ancora_runtime_new returned null");
            }
            return new RuntimeHandle(runtimePtr.reinterpret(Long.MAX_VALUE));
        }
    }

    /**
     * Return the raw native pointer for passing to FFI functions.
     *
     * @throws IllegalStateException if the handle has been closed
     */
    public MemorySegment rawPtr() {
        if (closed) throw new IllegalStateException("RuntimeHandle has been closed");
        return ptr;
    }

    @Override
    public void close() throws Throwable {
        if (!closed) {
            closed = true;
            AncoraNative.FREE_RUNTIME.invokeExact(ptr);
        }
    }
}

File: sdk/dotnet/Ancora/Handles/RuntimeHandle.cs
using System;
using System.Runtime.InteropServices;
using Ancora.Interop;

namespace Ancora.Handles;

/// <summary>
/// SafeHandle wrapping an opaque AncorRuntime pointer.
/// Calls ancora_free_runtime on finalization or explicit Dispose.
/// </summary>
internal sealed class RuntimeHandle : SafeHandle
{
    private RuntimeHandle() : base(IntPtr.Zero, ownsHandle: true) { }

    public override bool IsInvalid => handle == IntPtr.Zero;

    protected override bool ReleaseHandle()
    {
        AncoraNative.ancora_free_runtime(handle);
        return true;
    }

    internal static RuntimeHandle Create()
    {
        var rc = AncoraNative.ancora_runtime_new(out var ptr);
        if (rc != AncorErrorCode.Ok)
            throw new AncorException((int)rc, $"ancora_runtime_new returned {rc}");
        if (ptr == IntPtr.Zero)
            throw new AncorException((int)AncorErrorCode.Internal, "ancora_runtime_new returned null");
        var h = new RuntimeHandle();
        h.SetHandle(ptr);
        return h;
    }

    internal static unsafe RuntimeHandle Create(ReadOnlySpan<byte> configBytes)
    {
        AncorErrorCode rc;
        IntPtr ptr;
        fixed (byte* p = configBytes)
        {
            rc = AncoraNative.ancora_runtime_new_with_config(
                (IntPtr)p, (nuint)configBytes.Length, out ptr);
        }
        if (rc != AncorErrorCode.Ok)
            throw new AncorException((int)rc, $"ancora_runtime_new_with_config returned {rc}");
        if (ptr == IntPtr.Zero)
        {
            throw new AncorException(
                (int)AncorErrorCode.Internal, "ancora_runtime_new_with_config returned null");
        }
        var h = new RuntimeHandle();
        h.SetHandle(ptr);
        return h;
    }
}
